test/components/preprocessors/test_recursive_splitter.py

```diff
@@ -0,0 +1,210 @@
+import pytest
+from pytest import LogCaptureFixture
+
+from haystack import Document, Pipeline
+from haystack.components.preprocessors.recursive_splitter import RecursiveDocumentSplitter
+from haystack.components.preprocessors.sentence_tokenizer import SentenceSplitter
+
+
+def test_get_custom_sentence_tokenizer_success():
+    tokenizer = RecursiveDocumentSplitter._get_custom_sentence_tokenizer()
+    assert isinstance(tokenizer, SentenceSplitter)
+
+
+def test_init_with_negative_overlap():
+    with pytest.raises(ValueError):
+        _ = RecursiveDocumentSplitter(split_length=20, split_overlap=-1, separators=["."])
+
+
+def test_init_with_overlap_greater_than_chunk_size():
+    with pytest.raises(ValueError):
+        _ = RecursiveDocumentSplitter(split_length=10, split_overlap=15, separators=["."])
+
+
+def test_init_with_invalid_separators():
+    with pytest.raises(ValueError):
+        _ = RecursiveDocumentSplitter(separators=[".", 2])
+
+
+def test_init_with_negative_split_length():
+    with pytest.raises(ValueError):
+        _ = RecursiveDocumentSplitter(split_length=-1, separators=["."])
+
+
+def test_apply_overlap_no_overlap():
+    # Test the case where there is no overlap between chunks
+    splitter = RecursiveDocumentSplitter(split_length=20, split_overlap=0, separators=["."])
+    chunks = ["chunk1", "chunk2", "chunk3"]
+    result = splitter._apply_overlap(chunks)
+    assert result == ["chunk1", "chunk2", "chunk3"]
+
+
+def test_apply_overlap_with_overlap_case_1():
+    # Test the case where there is overlap between chunks
+    splitter = RecursiveDocumentSplitter(split_length=20, split_overlap=4, separators=["."])
+    chunks = ["chunk1", "chunk2", "chunk3"]
+    result = splitter._apply_overlap(chunks)
+    assert result == ["chunk1", "unk1chunk2", "unk2chunk3"]
+
+
+def test_apply_overlap_with_overlap_case_2():
+    # Test the case where there is overlap between chunks
+    splitter = RecursiveDocumentSplitter(split_length=20, split_overlap=6, separators=["."])
+    chunks = ["chunk1", "chunk2", "chunk3", "chunk4"]
+    result = splitter._apply_overlap(chunks)
+    assert result == ["chunk1", "chunk1chunk2", "chunk2chunk3", "chunk3chunk4"]
```

Comment:
So yeah this is the bug I want to make sure we avoid. Specifically, we want to avoid producing chunks that are full copies of previous chunks. So we shouldn't return both `chunk1` and `chunk1chunk2`. Preventing this was the original behavior in v1 and is the behavior of our NLTKDocumentSplitter when splitting by sentence.

Reply:
Looking more into the original fix for this, in this PR https://github.com/deepset-ai/haystack/pull/8539 I see that this might be a more involved fix, but in general we want to avoid fully copying previous chunks into new chunks where possible because this can negatively affect the performance of the rest of the pipeline.

Reply:
OK, now I understand what you wanted to test exactly. This can be handled with checks and validations, as you mentioned it was done on v1.x.

I can work on a fix for it, but it seems to me this is  more a responsibility of the caller, to understand it's data and the parameters of the splitter.

Reply:
I'd appreciate the fix on this! 

And I don't fully agree. As a caller it's really hard to fully know all of your data (especially tens of thousands of files). And so it's very possible in some scenarios you'll end up with chunks that are shorter than the split overlap and then this would occur. You're right it's theoretically possible someone could figure this out and try to strategize around it, but I like the idea of making the caller's life as easy as possible if we can.

Reply:
I will write the fix for this, no worries.

I was just trying to see this from your perspective - I see how this can become a big issue at large scale and very heterogeneous data.

Reply:
> Looking more into the original fix for this, in this PR https://github.com/deepset-ai/haystack/pull/8539 I see that this might be a more involved fix, but in general we want to avoid fully copying previous chunks into new chunks where possible because this can negatively affect the performance of the rest of the pipeline.

What would be the expected outcome to avoid a previous chunk being completely copied/overlapped into the current chunk?

- a warning saying that chunk from the previous document is being 100% copied to chunk from the current but continue
- raise an error and quit saying that chunk from the previous document is being 100% copied to chunk from the current but continue
- something else?